File: apps/web/src/lib/api/media.ts
export interface MediaFile {
  id: string
  libraryId: string
  filename: string
  sizeBytes: number
  durationMs: number
  width: number
  height: number
  container: string
  videoCodec: string
  audioCodec: string
  modifiedAt: string
  progressMs: number
  completed: boolean
  title: string
  description: string
  recordedAt: string | null
  favorite: boolean
}

export interface UpdateMediaInput {
  title: string
  description: string
  recordedAt: string | null
  favorite: boolean
}

export interface ScanResult {
  discovered: number
  indexed: number
  unchanged: number
  removed: number
  skipped: number
  issues: ScanIssue[]
}

export interface ScanIssue {
  filename: string
  reason: string
}

export interface ScanStatus {
  state: 'idle' | 'pending' | 'scanning' | 'completed' | 'failed'
  startedAt?: string
  finishedAt?: string
  result?: ScanResult
  lastError?: string
}

export interface PlaybackInfo {
  mode: 'direct' | 'hls'
  url: string
}

export interface HomeMedia {
  continueWatching: MediaFile[]
  recentlyAdded: MediaFile[]
}

export interface MediaSearchResult extends MediaFile {
  libraryName: string
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null
}

function isMediaFile(value: unknown): value is MediaFile {
  return (
    isRecord(value) &&
    typeof value.id === 'string' &&
    typeof value.libraryId === 'string' &&
    typeof value.filename === 'string' &&
    typeof value.sizeBytes === 'number' &&
    typeof value.durationMs === 'number' &&
    typeof value.width === 'number' &&
    typeof value.height === 'number' &&
    typeof value.container === 'string' &&
    typeof value.videoCodec === 'string' &&
    typeof value.audioCodec === 'string' &&
    typeof value.modifiedAt === 'string'
    && typeof value.progressMs === 'number'
    && typeof value.completed === 'boolean'
    && typeof value.title === 'string'
    && typeof value.description === 'string'
    && (value.recordedAt === null || typeof value.recordedAt === 'string')
    && typeof value.favorite === 'boolean'
  )
}

function isScanResult(value: unknown): value is ScanResult {
  return (
    isRecord(value) &&
    typeof value.discovered === 'number' &&
    typeof value.indexed === 'number' &&
    typeof value.unchanged === 'number' &&
    typeof value.removed === 'number' &&
    typeof value.skipped === 'number' &&
    Array.isArray(value.issues) &&
    value.issues.every((issue) => isRecord(issue) && typeof issue.filename === 'string' && typeof issue.reason === 'string')
  )
}

function isScanStatus(value: unknown): value is ScanStatus {
  return (
    isRecord(value) &&
    ['idle', 'pending', 'scanning', 'completed', 'failed'].includes(String(value.state)) &&
    (value.startedAt === undefined || typeof value.startedAt === 'string') &&
    (value.finishedAt === undefined || typeof value.finishedAt === 'string') &&
    (value.result === undefined || isScanResult(value.result)) &&
    (value.lastError === undefined || typeof value.lastError === 'string')
  )
}

function isPlaybackInfo(value: unknown): value is PlaybackInfo {
  return isRecord(value) && (value.mode === 'direct' || value.mode === 'hls') && typeof value.url === 'string'
}

function isHomeMedia(value: unknown): value is HomeMedia {
  return (
    isRecord(value) &&
    Array.isArray(value.continueWatching) &&
    value.continueWatching.every(isMediaFile) &&
    Array.isArray(value.recentlyAdded) &&
    value.recentlyAdded.every(isMediaFile)
  )
}

function isMediaSearchResult(value: unknown): value is MediaSearchResult {
  return isRecord(value) && isMediaFile(value) && typeof value.libraryName === 'string'
}

export async function getMedia(libraryId: string, signal?: AbortSignal): Promise<MediaFile[]> {
  const query = new URLSearchParams({ libraryId })
  const response = await apiFetch(`/api/media?${query.toString()}`, signal === undefined ? undefined : { signal })
  if (!response.ok) throw new Error('Impossible de charger les vidéos')

  const body: unknown = await response.json()
  if (!isRecord(body) || !Array.isArray(body.items) || !body.items.every(isMediaFile)) {
    throw new Error('La réponse des vidéos est invalide')
  }
  return body.items
}

export async function getFavorites(signal?: AbortSignal): Promise<MediaFile[]> {
  const response = await apiFetch('/api/favorites', signal === undefined ? undefined : { signal })
  if (!response.ok) throw new Error('Impossible de charger les favoris')
  const body: unknown = await response.json()
  if (!isRecord(body) || !Array.isArray(body.items) || !body.items.every(isMediaFile)) {
    throw new Error('La réponse des favoris est invalide')
  }
  return body.items
}

export async function getMediaByID(mediaId: string, signal?: AbortSignal): Promise<MediaFile> {
  const response = await apiFetch(`/api/media/${encodeURIComponent(mediaId)}`, signal === undefined ? undefined : { signal })
  if (!response.ok) throw new Error('Impossible de charger la vidéo')

  const body: unknown = await response.json()
  if (!isMediaFile(body)) throw new Error('La réponse de la vidéo est invalide')
  return body
}

export async function updateMedia(mediaId: string, input: UpdateMediaInput): Promise<MediaFile> {
  const response = await apiFetch(`/api/media/${encodeURIComponent(mediaId)}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(input),
  })
  if (!response.ok) throw new Error('Impossible de modifier la vidéo')
  const body: unknown = await response.json()
  if (!isMediaFile(body)) throw new Error('La réponse de modification est invalide')
  return body
}

export async function setMediaFavorite(mediaId: string, favorite: boolean): Promise<MediaFile> {
  const response = await apiFetch(`/api/media/${encodeURIComponent(mediaId)}/favorite`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ favorite }),
  })
  if (!response.ok) throw new Error('Impossible de modifier le favori')
  const body: unknown = await response.json()
  if (!isMediaFile(body)) throw new Error('La réponse vidéo est invalide')
  return body
}

export function thumbnailURL(mediaId: string): string {
  return `/api/media/${encodeURIComponent(mediaId)}/thumbnail`
}

export function streamURL(mediaId: string): string {
  return `/api/media/${encodeURIComponent(mediaId)}/stream`
}

export async function getPlayback(mediaId: string, signal?: AbortSignal): Promise<PlaybackInfo> {
  const response = await apiFetch(`/api/media/${encodeURIComponent(mediaId)}/playback`, signal === undefined ? undefined : { signal })
  if (!response.ok) throw new Error('Impossible de préparer la lecture')

  const body: unknown = await response.json()
  if (!isPlaybackInfo(body)) throw new Error('La réponse de lecture est invalide')
  return body
}

export async function getHomeMedia(signal?: AbortSignal): Promise<HomeMedia> {
  const response = await apiFetch('/api/home', signal === undefined ? undefined : { signal })
  if (!response.ok) throw new Error("Impossible de charger l'accueil")
  const body: unknown = await response.json()
  if (!isHomeMedia(body)) throw new Error("La réponse de l'accueil est invalide")
  return body
}

export async function searchMedia(query: string, signal?: AbortSignal): Promise<MediaSearchResult[]> {
  const parameters = new URLSearchParams({ q: query })
  const response = await apiFetch(`/api/search?${parameters.toString()}`, signal === undefined ? undefined : { signal })
  if (!response.ok) throw new Error('La recherche a échoué')
  const body: unknown = await response.json()
  if (!isRecord(body) || !Array.isArray(body.items) || !body.items.every(isMediaSearchResult)) {
    throw new Error('La réponse de recherche est invalide')
  }
  return body.items
}


export async function scanLibrary(libraryId: string): Promise<ScanResult> {
  const response = await apiFetch(`/api/libraries/${encodeURIComponent(libraryId)}/scan`, { method: 'POST' })
  if (!response.ok) throw new Error("L'analyse de la bibliothèque a échoué")

  const body: unknown = await response.json()
  if (!isScanResult(body)) throw new Error("La réponse d'analyse est invalide")
  return body
}

export async function getScanStatus(libraryId: string, signal?: AbortSignal): Promise<ScanStatus> {
  const response = await apiFetch(
    `/api/libraries/${encodeURIComponent(libraryId)}/scan`,
    signal === undefined ? undefined : { signal },
  )
  if (!response.ok) throw new Error("Impossible de connaître l'état de l'analyse")

  const body: unknown = await response.json()
  if (!isScanStatus(body)) throw new Error("La réponse d'état de l'analyse est invalide")
  return body
}
import { apiFetch } from './client'

File: apps/server/internal/httpserver/media.go
package httpserver

import (
	"encoding/json"
	"errors"
	"log/slog"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"

	"flex.local/server/internal/library"
	"flex.local/server/internal/media"
)

type mediaResponse struct {
	ID          string     `json:"id"`
	LibraryID   string     `json:"libraryId"`
	Filename    string     `json:"filename"`
	SizeBytes   int64      `json:"sizeBytes"`
	DurationMS  int64      `json:"durationMs"`
	Width       int        `json:"width"`
	Height      int        `json:"height"`
	Container   string     `json:"container"`
	VideoCodec  string     `json:"videoCodec"`
	AudioCodec  string     `json:"audioCodec"`
	ModifiedAt  time.Time  `json:"modifiedAt"`
	ProgressMS  int64      `json:"progressMs"`
	Completed   bool       `json:"completed"`
	Title       string     `json:"title"`
	Description string     `json:"description"`
	RecordedAt  *time.Time `json:"recordedAt"`
	Favorite    bool       `json:"favorite"`
}

type updateMediaRequest struct {
	Title       string  `json:"title"`
	Description string  `json:"description"`
	RecordedAt  *string `json:"recordedAt"`
	Favorite    bool    `json:"favorite"`
}

type mediaListResponse struct {
	Items []mediaResponse `json:"items"`
}

type mediaSearchResponse struct {
	Items []mediaSearchItemResponse `json:"items"`
}

type mediaSearchItemResponse struct {
	mediaResponse
	LibraryName string `json:"libraryName"`
}

type homeResponse struct {
	ContinueWatching []mediaResponse `json:"continueWatching"`
	RecentlyAdded    []mediaResponse `json:"recentlyAdded"`
}

type scanResponse struct {
	Discovered int `json:"discovered"`
	Indexed    int `json:"indexed"`
	Skipped    int `json:"skipped"`
}

type playbackResponse struct {
	Mode media.PlaybackMode `json:"mode"`
	URL  string             `json:"url"`
}

type folderAssignmentResponse struct {
	MediaID string `json:"mediaId"`
	Folder  string `json:"folder"`
}

type folderAssignmentListResponse struct {
	Items []folderAssignmentResponse `json:"items"`
}

func libraryFoldersHandler(service mediaService, logger *slog.Logger) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {
		items, err := service.Folders(request.Context(), request.PathValue("libraryID"))
		if err != nil {
			logger.Error("list library folders", "error", err)
			writeError(response, http.StatusInternalServerError, "internal_error", "Impossible de charger les dossiers")
			return
		}
		payload := folderAssignmentListResponse{Items: make([]folderAssignmentResponse, 0, len(items))}
		for _, item := range items {
			payload.Items = append(payload.Items, folderAssignmentResponse{MediaID: item.MediaID, Folder: item.Folder})
		}
		response.Header().Set("Content-Type", "application/json; charset=utf-8")
		_ = json.NewEncoder(response).Encode(payload)
	}
}

func listMediaHandler(service mediaService, logger *slog.Logger) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {
		libraryID := request.URL.Query().Get("libraryId")
		if libraryID == "" {
			writeError(response, http.StatusBadRequest, "missing_library", "L'identifiant de la bibliothèque est requis")
			return
		}
		items, err := service.List(request.Context(), libraryID)
		if err != nil {
			logger.Error("list media", "error", err)
			writeError(response, http.StatusInternalServerError, "internal_error", "Impossible de charger les vidéos")
			return
		}
		payload := mediaListResponse{Items: make([]mediaResponse, 0, len(items))}
		for _, item := range items {
			payload.Items = append(payload.Items, mediaToResponse(item))
		}
		response.Header().Set("Content-Type", "application/json; charset=utf-8")
		_ = json.NewEncoder(response).Encode(payload)
	}
}

func favoritesHandler(service mediaService, logger *slog.Logger) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {
		items, err := service.Favorites(request.Context())
		if err != nil {
			logger.Error("list favorite media", "error", err)
			writeError(response, http.StatusInternalServerError, "internal_error", "Impossible de charger les favoris")
			return
		}
		payload := mediaListResponse{Items: make([]mediaResponse, 0, len(items))}
		for _, item := range items {
			payload.Items = append(payload.Items, mediaToResponse(item))
		}
		response.Header().Set("Content-Type", "application/json; charset=utf-8")
		_ = json.NewEncoder(response).Encode(payload)
	}
}

func searchMediaHandler(service mediaService, logger *slog.Logger) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {
		items, err := service.Search(request.Context(), request.URL.Query().Get("q"))
		if err != nil {
			logger.Error("search media", "error", err)
			writeError(response, http.StatusInternalServerError, "internal_error", "La recherche a échoué")
			return
		}
		payload := mediaSearchResponse{Items: make([]mediaSearchItemResponse, 0, len(items))}
		for _, item := range items {
			payload.Items = append(payload.Items, mediaSearchItemResponse{mediaResponse: mediaToResponse(item.File), LibraryName: item.LibraryName})
		}
		response.Header().Set("Content-Type", "application/json; charset=utf-8")
		_ = json.NewEncoder(response).Encode(payload)
	}
}

func homeHandler(service mediaService, logger *slog.Logger) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {
		home, err := service.Home(request.Context())
		if err != nil {
			logger.Error("load home", "error", err)
			writeError(response, http.StatusInternalServerError, "internal_error", "Impossible de charger l'accueil")
			return
		}
		payload := homeResponse{
			ContinueWatching: make([]mediaResponse, 0, len(home.ContinueWatching)),
			RecentlyAdded:    make([]mediaResponse, 0, len(home.RecentlyAdded)),
		}
		for _, item := range home.ContinueWatching {
			payload.ContinueWatching = append(payload.ContinueWatching, mediaToResponse(item))
		}
		for _, item := range home.RecentlyAdded {
			payload.RecentlyAdded = append(payload.RecentlyAdded, mediaToResponse(item))
		}
		response.Header().Set("Content-Type", "application/json; charset=utf-8")
		_ = json.NewEncoder(response).Encode(payload)
	}
}

func getMediaHandler(service mediaService, logger *slog.Logger) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {
		item, err := service.Get(request.Context(), request.PathValue("mediaID"))
		if errors.Is(err, media.ErrNotFound) {
			writeError(response, http.StatusNotFound, "media_not_found", "La vidéo n'existe pas")
			return
		}
		if err != nil {
			logger.Error("get media", "error", err)
			writeError(response, http.StatusInternalServerError, "internal_error", "Impossible de charger la vidéo")
			return
		}
		response.Header().Set("Content-Type", "application/json; charset=utf-8")
		_ = json.NewEncoder(response).Encode(mediaToResponse(item))
	}
}

func updateMediaHandler(service mediaService, logger *slog.Logger) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {
		var input updateMediaRequest
		decoder := json.NewDecoder(http.MaxBytesReader(response, request.Body, 1<<20))
		decoder.DisallowUnknownFields()
		if err := decoder.Decode(&input); err != nil {
			writeError(response, http.StatusBadRequest, "invalid_request", "La requête est invalide")
			return
		}
		var recordedAt *time.Time
		if input.RecordedAt != nil && *input.RecordedAt != "" {
			parsed, err := time.Parse("2006-01-02", *input.RecordedAt)
			if err != nil {
				writeError(response, http.StatusUnprocessableEntity, "invalid_recorded_at", "La date d'enregistrement est invalide")
				return
			}
			recordedAt = &parsed
		}
		updated, err := service.UpdateMetadata(request.Context(), request.PathValue("mediaID"), media.MetadataInput{
			Title: input.Title, Description: input.Description, RecordedAt: recordedAt, Favorite: input.Favorite,
		})
		switch {
		case errors.Is(err, media.ErrNotFound):
			writeError(response, http.StatusNotFound, "media_not_found", "La vidéo n'existe pas")
		case errors.Is(err, media.ErrInvalidTitle):
			writeError(response, http.StatusUnprocessableEntity, "invalid_title", "Le titre est requis et doit contenir moins de 200 caractères")
		case err != nil:
			logger.Error("update media metadata", "error", err)
			writeError(response, http.StatusInternalServerError, "internal_error", "Impossible de modifier la vidéo")
		default:
			response.Header().Set("Content-Type", "application/json; charset=utf-8")
			_ = json.NewEncoder(response).Encode(mediaToResponse(updated))
		}
	}
}

func thumbnailHandler(service mediaService, logger *slog.Logger) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {
		path, err := service.Thumbnail(request.Context(), request.PathValue("mediaID"))
		if errors.Is(err, media.ErrNotFound) {
			http.NotFound(response, request)
			return
		}
		if err != nil {
			logger.Error("generate thumbnail", "error", err)
			http.Error(response, "thumbnail unavailable", http.StatusInternalServerError)
			return
		}
		response.Header().Set("Cache-Control", "public, max-age=86400")
		http.ServeFile(response, request, path)
	}
}

func streamHandler(service mediaService, logger *slog.Logger) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {
		item, err := service.Get(request.Context(), request.PathValue("mediaID"))
		if errors.Is(err, media.ErrNotFound) {
			http.NotFound(response, request)
			return
		}
		if err != nil {
			logger.Error("get media stream", "error", err)
			http.Error(response, "stream unavailable", http.StatusInternalServerError)
			return
		}
		file, err := os.Open(item.Path)
		if err != nil {
			logger.Error("open media stream", "error", err)
			http.Error(response, "stream unavailable", http.StatusInternalServerError)
			return
		}
		defer file.Close()
		info, err := file.Stat()
		if err != nil {
			http.Error(response, "stream unavailable", http.StatusInternalServerError)
			return
		}
		http.ServeContent(response, request, item.Filename, info.ModTime(), file)
	}
}

func playbackHandler(service mediaService, logger *slog.Logger) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {
		mediaID := request.PathValue("mediaID")
		item, err := service.Get(request.Context(), mediaID)
		if errors.Is(err, media.ErrNotFound) {
			writeError(response, http.StatusNotFound, "media_not_found", "La vidéo n'existe pas")
			return
		}
		if err != nil {
			logger.Error("get playback mode", "error", err)
			writeError(response, http.StatusInternalServerError, "internal_error", "Impossible de préparer la lecture")
			return
		}
		mode := media.SelectPlaybackMode(item)
		url := "/api/media/" + mediaID + "/stream"
		if mode == media.PlaybackHLS {
			url = "/api/media/" + mediaID + "/hls/index.m3u8"
		}
		response.Header().Set("Content-Type", "application/json; charset=utf-8")
		_ = json.NewEncoder(response).Encode(playbackResponse{Mode: mode, URL: url})
	}
}

func hlsHandler(service mediaService, logger *slog.Logger) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {
		filename := request.PathValue("filename")
		if !validHLSFilename(filename) {
			http.NotFound(response, request)
			return
		}
		directory, err := service.Transcode(request.Context(), request.PathValue("mediaID"))
		if errors.Is(err, media.ErrNotFound) {
			http.NotFound(response, request)
			return
		}
		if err != nil {
			logger.Error("transcode media", "error", err)
			http.Error(response, "transcode unavailable", http.StatusInternalServerError)
			return
		}
		if filename == "index.m3u8" {
			response.Header().Set("Content-Type", "application/vnd.apple.mpegurl")
			response.Header().Set("Cache-Control", "no-cache")
		} else {
			response.Header().Set("Content-Type", "video/mp2t")
			response.Header().Set("Cache-Control", "public, max-age=86400")
		}
		http.ServeFile(response, request, filepath.Join(directory, filename))
	}
}

func validHLSFilename(filename string) bool {
	if filename == "index.m3u8" {
		return true
	}
	if !strings.HasPrefix(filename, "segment-") || !strings.HasSuffix(filename, ".ts") {
		return false
	}
	digits := strings.TrimSuffix(strings.TrimPrefix(filename, "segment-"), ".ts")
	if digits == "" {
		return false
	}
	for _, character := range digits {
		if character < '0' || character > '9' {
			return false
		}
	}
	return true
}

func scanLibraryHandler(service scanService, logger *slog.Logger) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {
		result, err := service.Scan(request.Context(), request.PathValue("libraryID"))
		if errors.Is(err, library.ErrNotFound) {
			writeError(response, http.StatusNotFound, "library_not_found", "La bibliothèque n'existe pas")
			return
		}
		if err != nil {
			logger.Error("scan library", "error", err)
			writeError(response, http.StatusInternalServerError, "scan_failed", "L'analyse de la bibliothèque a échoué")
			return
		}
		response.Header().Set("Content-Type", "application/json; charset=utf-8")
		_ = json.NewEncoder(response).Encode(scanResponse{
			Discovered: result.Discovered,
			Indexed:    result.Indexed,
			Skipped:    result.Skipped,
		})
	}
}

func scanStatusHandler(service scanService) http.HandlerFunc {
	return func(response http.ResponseWriter, request *http.Request) {
		status := service.Status(request.PathValue("libraryID"))
		response.Header().Set("Content-Type", "application/json; charset=utf-8")
		_ = json.NewEncoder(response).Encode(struct {
			State     string     `json:"state"`
			StartedAt *time.Time `json:"startedAt,omitempty"`
			LastError string     `json:"lastError,omitempty"`
		}{State: status.State, StartedAt: status.StartedAt, LastError: status.LastError})
	}
}

func mediaToResponse(item media.File) mediaResponse {
	return mediaResponse{
		ID: item.ID, LibraryID: item.LibraryID, Filename: item.Filename, SizeBytes: item.SizeBytes,
		DurationMS: item.DurationMS, Width: item.Width, Height: item.Height, Container: item.Container,
		VideoCodec: item.VideoCodec, AudioCodec: item.AudioCodec, ModifiedAt: item.ModifiedAt,
		ProgressMS: item.ProgressMS, Completed: item.Completed,
		Title: item.Title, Description: item.Description, RecordedAt: item.RecordedAt, Favorite: item.Favorite,
	}
}
